lua/core/keyboard.lua

```diff
@@ -0,0 +1,207 @@
+keyboard = {}
+
+keyboard.keymap = {}
+keyboard.keymap.us = require 'core/keymap/us' 
+keyboard.selected_map = "us"
+
+local km = keyboard.keymap[keyboard.selected_map]
+
+
+keyboard.state = {
+  Lshift = false,
+  Rshift = false,
+  Lalt = false,
+  Ralt = false,
+  Lctrl = false,
+  Rctrl = false,
+  Lmeta = false,
+  Rmeta = false
+}
+
+function keyboard.clear()
+  keyboard.code = nil      -- callback for code keyboard codes
+  keyboard.ascii = nil    -- callback for ascii keycodes, by map
+end
+
+function keyboard.set_map(m)
+  if keyboard.keymap[m] then
+    keyboard.selected_map = m
+  end
+end
+
+function keyboard.shift()
+  return keyboard.state.Lshift or keyboard.state.Rshift end
+function keyboard.alt()
+  return keyboard.state.Lalt or keyboard.state.Ralt end
+function keyboard.ctrl()
+  return keyboard.state.Lctrl or keyboard.state.Rctrl end
+function keyboard.meta()
+  return keyboard.state.Lmeta or keyboard.state.Rmeta end
+
+
+function keyboard.process(type,code,value)
+  if keyboard.code then
+    keyboard.code(keyboard.codes[code],value)
+  end
+
+  local c = keyboard.codes[code] 
+  if keyboard.process_mods[c] then
+    keyboard.process_mods[c](value)
+  end
+
+  if value>0 then
+    local a = km[keyboard.shift()][c]
+    if a then 
+      if keyboard.ascii then
+        keyboard.ascii(a)
+      end
+      --print("ASCII: "..a)
+    end
+  end
+
+  --print("kb",code,value,keyboard.codes[code])
+end
+
+keyboard.process_mods = {}
+keyboard.process_mods.LEFTSHIFT = function(value) keyboard.state.Lshift = (value>0) end
+keyboard.process_mods.RIGHTSHIFT = function(value) keyboard.state.Rshift = (value>0) end
+keyboard.process_mods.LEFTALT = function(value) keyboard.state.Lalt = (value>0) end
+keyboard.process_mods.RIGHTALT = function(value) keyboard.state.Ralt = (value>0) end
+keyboard.process_mods.LEFTCTRL = function(value) keyboard.state.Lctrl = (value>0) end
+keyboard.process_mods.RIGHTSHIFT = function(value) keyboard.state.Rctrl = (value>0) end
+keyboard.process_mods.LEFTMETA = function(value) keyboard.state.Lmeta = (value>0) end
+keyboard.process_mods.RIGHTMETA= function(value) keyboard.state.Rmeta = (value>0) end
+
```

Comment:
this also seems maybe unnecessarily repetitive. like, if the modifier flags in `keyboard.state` shared the same keys as the relevant keycodes , this could be one function `keyboard.process_mod = function(val) keyboard.state[val] = val > 0 end`